aim/agent/aid/universes/aci/aci_universe.py

```diff
@@ -113,37 +124,148 @@ def _reload_websocket_config(self):
             self.ws_urls = ws_urls
 
+    def _update_apic_assign_db(self, aim_context, apic_assign,
+                               apic_assign_obj):
+        try:
+            if apic_assign_obj is None:
+                apic_assign.aim_aid_id = self.agent_id
+                return self.manager.create(aim_context, apic_assign)
+
+            if apic_assign_obj.aim_aid_id != self.agent_id:
+                obj = self.manager.update(aim_context, apic_assign_obj,
+                                          aim_aid_id=self.agent_id)
+            else:
+                # This will update the last_update_timestamp
+                # automatically
+                obj = self.manager.update(aim_context, apic_assign_obj)
+            return obj
+        # This means another controller is also adding/updating
+        # the same entry at the same time and he has beat us.
+        except Exception as e:
+            LOG.info(e)
+            return None
+
+    def _ws_session_login(self, url, url_max_reties, purpose):
+        retries = 0
+        LOG.info('Establishing %s WS connection with url: %s',
+                 purpose, url)
+        while retries < url_max_reties:
+            session = acitoolkit.Session(
+                url, self.apic_username,
+                self.apic_password,
+                verify_ssl=self.verify_ssl_certificate,
+                cert_name=self.cert_name,
+                key=self.private_key_file)
+            resp = session.login()
+            if not resp.ok:
+                LOG.warn('%s Websocket connection failed: %s',
+                         purpose, resp.text)
+                retries += 1
+                if session.session:
+                    session.close()
+                continue
+            LOG.info('%s Websocket connection succeeded with url: %s',
+                     purpose, url)
+            return session
+
+        return None
+
+    def establish_ws_session_for_recovery(self, max_retries=None):
+        try:
+            with utils.get_rlock(lcon.ACI_WS_CONNECTION_LOCK, blocking=False):
+                max_retries = max_retries or 2 * len(self.ws_urls)
+                url_max_reties = max(1, max_retries / len(self.ws_urls))
+                aim_context = aim_ctx.AimContext(store=api.get_store())
+                for url in self.ws_urls:
+                    apic_assign = api_infra.ApicAssignment(apic_host=url)
+                    apic_assign_obj = self.manager.get(aim_context,
+                                                       apic_assign)
+                    if (apic_assign_obj and
+                        apic_assign_obj.aim_aid_id != self.agent_id and
+                            not apic_assign_obj.is_available(aim_context)):
+                        continue
+                    # This means the original aim-aid owner might have
+                    # crashed or something. We will just take it!
+                    if apic_assign_obj and self.session.ipaddr in url:
```

Comment:
good point. Let me think about it then. Worst case I just don't do this special thing. Then it will always try to re-connect.

Reply:
By examining the code, actually this will be ok coz self.session.ipaddr could also be DNS name. The code basically just strips out the "https://" from the beginning then assign it to ipaddr member variable. Agreed that ipaddr naming is confusing somehow.....
https://github.com/noironetworks/acitoolkit/blob/noiro-lite/acitoolkit/acisession.py#L503-L506

Reply:
ah, good point. That works then.